### ISSUE TITLE
`super().__init__()` in BaseModule breaks inheritance for other projects

### ISSUE THREAD

**Issue by watsonjj** [tsfpga/tsfpga#112]
In our project which depends on tsfpga, we have the a class which extends the `BaseModule`:

```python
class CorelModule(BaseModule, StackModule):
```

We then initialise the `BaseModule` first in `CorelModule.__init__`, and then later the `StackModule`.

```python
def __init__(
            self,
            ...
    ):
        BaseModule.__init__(
            self,
            path=path,
            library_name=library_name,
            default_registers=default_registers
        )
...
        StackModule.__init__(
            self,
            instance=instance,
            parent=parent,
            address_space_size=address_space_size
        )
```

Since the change in #84 / #85, multiple inheritance involving `BaseModule` in external packages could be broken, as the call to `StackModule.__init__` now occurs during `BaseModule.__init__`, but without the necessary arguments.

We can fix our error by changing the inheritance order:

```python
class CorelModule(StackModule, BaseModule):
```

Which is the same solution highlighted in #84, and points out that the solution implemented in #85 is not ideal. I wonder if the following solution to the original problem in #84 would have been feasible, instead of complicating multiple inheritance downstream:

```python
class Module(BaseModuleLocal, FuncMixin):
    def __init__(self, *args):
        BaseModuleLocal.__init__(self, *args)
        FuncMixin.__init__(self)
```

**Comment by LukasVik** [comment]
I have never really had this use case, and I'm a little bit out of my element, so I personally don't have an opinion at this point.

@ru551n and @watsonjj, do you think you can land on a solution that you are both happy with?

**Comment by ru551n** [comment]
With that said, it is probably a better pattern to super().__init__ first since you can override members in the parent class that way.

**Comment by watsonjj** [comment]
> each module in the MRO must have a call to super().__init__ last in it's __init__ to be independent of the MRO. That is what https://github.com/tsfpga/tsfpga/pull/85 fixes in the first place. The call to super().__init__ simply calls the next module in the MRO, not necessarily the parent class.

This is only one approach to multiple inheritance in Python, where you are at the whims of the MRO (not independent from it at all). If you rely on this approach, and both of your parent classes require some arguments to the `__init__`, then these args can not be easily passed to the other class unless the first class in the MRO accepts `*args/**kwargs` for this purpose. In my opinion, Python classes should not need to be designed with the foresight that they will be used in inheritance.

The other standard approach for multiple inheritance is the one I demonstrated, where you call the `__init__` of the parent classes in turn. This does not require any special design of the parent classes to be able to function.

In our particular use case for the `CorelModule` class, we init the `BaseModule`, do some work to build the args of `StackModule` based on the registers in the `BaseModule`, then init the `StackModule`. This cannot work in your approach.

**Comment by ru551n** [comment]
Got it, (TIL!) I did not realize that there were dependencies in your `__init__` calls, my bad, sorry!

Is there any particular reason why it would not work in your case? If you could elaborate a bit about `...` in between your `__init__` calls, is there any specific stuff in between there that e.g could not be placed in `StackModule`, or even an intermediate class of sort? I feel that there might be some design pattern that does not align here. 

In our case, we have one `CustomBaseModule` which inherits directly from `BaseModule` anda whole bunch of mixins adding functionality, each `module_xx.py` uses only `CustomBaseModule` and the desired mixins.

Reverting this change would break our current code base and will trigger quite a lot of rework. Ultimately I think it is up to @LukasVik to decide what approach to go for since it is ultimately a question of a design choice.

**Comment by LukasVik** [comment]
> the call to `StackModule.__init__` now occurs during `BaseModule.__init__`, but **without the necessary arguments**.

This is something that I missed when merging #85. Sorry.

To me it seems absolutely clear that we must support inheritance from multiple classes to `BaseModule`. It also seems very reasonable that some of those base class constructors might require further arguments set by `BaseModule.__init__`. I have never had that use case but it seems reasonable.

Am I understanding it correctly that this is not possible after #85? 

Googling the Python class constructor MRO was quite overwhelming. I think it is good if we can save our users that headache. Even though the constructor execution order seems deterministic and well-defined, just like PEP20 (https://peps.python.org/pep-0020/#the-zen-of-python), I prefer explicit over implicit. If I was writing a child class with multiple base constructors, I think I would want to write the call order manually in my child constructor. At least if I do something at least remotely exotic.

To me, right now, the best solution seems to be to roll back #85. And instead force users write their own explicit constructor calls. E.g.

```
class Module(BaseModuleLocal, FuncMixin):
    def __init__(self, ...):
        BaseModule.__init__(...)
        StackModule.__init__(...)
```

```
class Module(BaseModuleLocal, FuncMixin):
    def __init__(self, ...):
        BaseModuleLocal.__init__(...)
        FuncMixin.__init__(...)
```

Any thoughts? I am open for all input.

**Comment by ru551n** [comment]
I agree. After considering the uses cases I believe it is a reasonable to have users do the explicit initialization. Explicit over implicit is the correct way to go.

I will forward this request and plan it accordingly. I have already done some refactoring and seems doable without too much of an effort.

**Comment by LukasVik** [comment]
#85 was included in release 12.3.4 (9 july 2024), which was a patch release (13.3.3 -> 12.3.4). I did not realize that this change would be breaking (it is not even included in the release notes). 

I think when we revert this, it has to be considered an API breaking change, resulting in release 13.0.0. Agree?

**Comment by LukasVik** [comment]
Fixed in PR #115

**Comment by LukasVik** [comment]
The fix is now released in version 13.0.0
